File: 1999/5/6-5.py
#!/usr/bin/env python

import sys

train = 1
for line in sys.stdin.read().splitlines():
    ln = map(int, line.split())
    cnt = ln[0]
    num = ln[1]
    incoming = ln[2:]
    holds = []
    for i in xrange(num):
        holds.append([])
    out = 0
    while len(incoming) > 0:
        cur = incoming.pop()
        if cur - 1 == out:
            out = cur
            move = 1
            while move:
                move = 0
                for h in holds:
                    if len(h) != 0:
                        if h[-1] - 1 == out:
                            move = 1
                            out = h.pop()
        else:
            i = 0
            while i < len(holds):
                tmp = holds[i]
                if len(tmp) == 0:
                    tmp.append(cur)
                    holds[i] = tmp
                    break
                elif cur < tmp[-1]:
                    tmp.append(cur)
                    holds[i] = tmp
                    break
                i += 1
            if i == len(holds):
                holds.append([cur])
    if len(holds) == num:
        print "Train", train, "can be rearranged."
    else:
        print "Train", train, "cannot be rearranged. At least",
        print len(holds), "holding tracks would be needed."
    train += 1

File: 1999/5/6-5.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <stack>
using namespace std;

int main()
{
	ifstream in("prob5_in.txt");
	stack<int> incoming, newstack;
	vector< stack<int> > holds;
	int num, cnt, train = 0, tmp, out;
	while (in >> cnt)
	{
		train++;
		in >> num;
		for (int i = 0; i < cnt; i++)
		{
			in >> tmp;
			incoming.push(tmp);
		}
		holds.reserve(num);
		out = 0;
		while (incoming.size() > 0)
		{
			char cur = incoming.top();
			incoming.pop();
			if (cur - 1 == out)
			{
				out = cur;
				bool move;
				do
				{
					move = false;
					for (int i = 0; i < holds.size(); i++)
					{
						if (holds[i].top() - 1 == out)
						{
							move = true;
							out = holds[i].top();
							holds[i].pop();
						}
					}
				}
				while (move);
			}
			else
			{
				for (int i = 0; i < holds.size(); i++)
				{
					if (holds[i].empty())
					{
						holds[i].push(cur);
						break;
					}
					else if (cur < holds[i].top())
					{
						holds[i].push(cur);
						break;
					}
				}
				if (i == holds.size())
				{
					holds.push_back(newstack);
					holds[i].push(cur);
				}
			}
		}
		if (holds.size() == num)
			cout << "Train " << train << " can be rearranged." << endl;
		else
			cout << "Train " << train << " cannot be rearranged. At least "
				<< holds.size() << " holding tracks would be needed." << endl;
		holds.clear();
	}
}
